File: android/app/src/main/java/com/bayesmech/vision/sensors/SensorSnapshot.kt
package com.bayesmech.vision.sensors

import com.bayesmech.vision.GpsLocation
import com.bayesmech.vision.ImuData
import com.bayesmech.vision.Vector3

data class SensorSnapshot(
    val angularVelocity: FloatArray,
    val magneticField: FloatArray,
    val gravity: FloatArray,
    val linearAcceleration: FloatArray,
    val gpsLocation: GpsLocation?
) {
    fun toImuData(): ImuData {
        val builder = ImuData.newBuilder()
        vectorOrNull(angularVelocity)?.let { builder.angularVelocity = it }
        vectorOrNull(linearAcceleration)?.let { builder.linearAcceleration = it }
        vectorOrNull(gravity)?.let { builder.gravity = it }
        vectorOrNull(magneticField)?.let { builder.magneticField = it }
        return builder.build()
    }

    fun summary(): String {
        val gpsStr = gpsLocation?.let {
            "GPS: %.6f, %.6f".format(it.latitude, it.longitude)
        } ?: "GPS: N/A"
        return "Accel: [%.2f, %.2f, %.2f] m/s², Gyro: [%.2f, %.2f, %.2f] rad/s, Gravity: [%.2f, %.2f, %.2f] m/s², %s".format(
            linearAcceleration[0], linearAcceleration[1], linearAcceleration[2],
            angularVelocity[0], angularVelocity[1], angularVelocity[2],
            gravity[0], gravity[1], gravity[2],
            gpsStr
        )
    }

    private fun vectorOrNull(values: FloatArray): Vector3? {
        if (values.none { it != 0f }) {
            return null
        }
        return Vector3.newBuilder()
            .setX(values[0])
            .setY(values[1])
            .setZ(values[2])
            .build()
    }
}

File: ios/BayesMechVision/Sensors/SensorSnapshot.swift
import Foundation
import simd

struct SensorSnapshot: Sendable {
    var angularVelocity: SIMD3<Float> = .zero
    var magneticField: SIMD3<Float> = .zero
    var gravity: SIMD3<Float> = .zero
    var linearAcceleration: SIMD3<Float> = .zero
    var gpsLocation: GpsLocation?

    func toImuData() -> ImuData? {
        var imu = ImuData()
        var populated = false

        if let angularVelocity = vectorOrNil(angularVelocity) {
            imu.angularVelocity = angularVelocity
            populated = true
        }
        if let linearAcceleration = vectorOrNil(linearAcceleration) {
            imu.linearAcceleration = linearAcceleration
            populated = true
        }
        if let gravity = vectorOrNil(gravity) {
            imu.gravity = gravity
            populated = true
        }
        if let magneticField = vectorOrNil(magneticField) {
            imu.magneticField = magneticField
            populated = true
        }

        return populated ? imu : nil
    }

    func summary() -> String {
        let gps = gpsLocation.map { String(format: "GPS: %.6f, %.6f", $0.latitude, $0.longitude) } ?? "GPS: N/A"
        return String(
            format: "Accel: [%.2f, %.2f, %.2f] m/s², Gyro: [%.2f, %.2f, %.2f] rad/s, Gravity: [%.2f, %.2f, %.2f] m/s², %@",
            linearAcceleration.x, linearAcceleration.y, linearAcceleration.z,
            angularVelocity.x, angularVelocity.y, angularVelocity.z,
            gravity.x, gravity.y, gravity.z,
            gps
        )
    }

    private func vectorOrNil(_ value: SIMD3<Float>) -> Vector3? {
        guard value != .zero else { return nil }
        var vector = Vector3()
        vector.x = value.x
        vector.y = value.y
        vector.z = value.z
        return vector
    }
}
